braidpool-primitives/src/bead/mod.rs

```diff
@@ -1,50 +1,182 @@
-#![allow(unused)]
-// Standard Imports
+use crate::braid::Braid;
+use crate::utils::{
+    BeadHash, BeadLoadError, Parents, hashset_to_vec_deterministic, vec_to_hashset,
+};
 use ::serde::{Deserialize, Serialize};
-use std::collections::HashSet;
-use std::net::SocketAddr;
-
-// Bitcoin primitives
-use bitcoin::Address;
+use bitcoin::BlockHash;
+use bitcoin::CompactTarget;
+use bitcoin::PublicKey;
 use bitcoin::absolute::Time;
+use bitcoin::consensus::Error;
+use bitcoin::consensus::encode::Decodable;
+use bitcoin::consensus::encode::Encodable;
 use bitcoin::ecdsa::Signature;
+use bitcoin::io::{self, BufRead, Write};
 use bitcoin::p2p::Address as P2P_Address;
-use bitcoin::secp256k1::PublicKey;
-use bitcoin::transaction::TransactionExt;
+use bitcoin::p2p::address::AddrV2;
 use bitcoin::{BlockHeader, Transaction};
-// Custom Imports
-use crate::utils::BeadHash;
-#[derive(Clone, Debug, Serialize)]
+use core::str::FromStr;
+use std::collections::HashSet;
+
+#[derive(Clone, Debug, Serialize, Deserialize, PartialEq, Eq, Default)]
+pub struct TimeVec(pub Vec<Time>);
+
+impl Encodable for TimeVec {
+    fn consensus_encode<W: Write + ?Sized>(&self, w: &mut W) -> Result<usize, io::Error> {
+        let mut len = 0;
+        // Encode the length for deterministic encoding
+        len += (self.0.len() as u64).consensus_encode(w)?;
+        for time in &self.0 {
+            len += time.to_consensus_u32().consensus_encode(w)?;
+        }
+        Ok(len)
+    }
+}
+
+impl Decodable for TimeVec {
+    fn consensus_decode<R: BufRead + ?Sized>(r: &mut R) -> Result<Self, Error> {
+        let len = u64::consensus_decode(r)?;
+        let mut vec = Vec::with_capacity(len as usize);
+        for _ in 0..len {
+            let time_u32 = u32::consensus_decode(r)?;
+            let time = Time::from_consensus(time_u32).unwrap();
+            vec.push(time);
+        }
+        Ok(TimeVec(vec))
+    }
+}
+
+#[derive(Clone, Debug, PartialEq)]
 
 pub struct CommittedMetadata {
-    // Committed Braidpool Metadata,
     pub transaction_cnt: u32,
     pub transactions: Vec<Transaction>,
     pub parents: HashSet<BeadHash>,
-    pub payout_address: Address,
-    //timestamp when the bead was created
-    pub observed_time_at_node: Time,
+    pub payout_address: P2P_Address,
+    pub start_timestamp: Time,
     pub comm_pub_key: PublicKey,
-    pub miner_ip: SocketAddr,
+    //minimum possible target > which will be the weak target
+    pub min_target: CompactTarget,
+    //the weaker target locallay apart from mainnet target ranging between the mainnet target and
```

Comment:
nit
or maybe rewrite a bit, the sentence is not so easy to understand
```suggestion
    //the weaker target locally apart from mainnet target ranging between the mainnet target and
```